### PR TITLE
Slightly reorganising the test suite to follow its documentation

### LINKED ISSUE #10902: Extraction to JSON is not documented.
#### Description of the problem

I discovered in #10897 that there was a JSON backend for extraction. I didn't know that and it is not documented anywhere I can see. In particular, the [`Extraction Language`](https://coq.github.io/doc/master/refman/addendum/extraction.html#coq:cmd.extraction-language) command is documented as only taking a value among `OCaml`, `Haskell`, and `Scheme`.

#### Coq Version

8.11 (current master) and below

### PR DESCRIPTION
**Pull request by Mbodin**
**Kind:** documentation / infrastructure.

This is a split of PR #12821 containing only the parts touching the documentation and the test suite.

Fixes #10902.

### CODE REVIEW

**Comment by Mbodin** [comment]
I feel that I shouldn’t squash the first commit with the others: it fixes a separate issue and only touches documentation. Maybe leaving the second commit unsquashed is fine too as it only adds a test and does not reorganise anything?

I think that it is safe to squash the last two, adding @herbelin as co-author. Would that be fine?

**Comment by Mbodin** [comment]
(Sorry for the multiple forced-push: I kept getting the wrong commit message for the first commit.)

**Comment by SkySkimmer** [comment]
The test for #5996 is wrong btw (compare with the examle in the issue)

**Comment by Mbodin** [comment]
Good catch! I didn’t change it: I just moved it so that it would be tested by the continuous integration system (it is currently ignored by it).

The current test is as follows:
```coq
Goal Type.
  let c := constr:(prod nat nat) in
  let c' := (eval pattern nat in c) in
  let c' := lazymatch c' with ?f _ => f end in
  let c'' := lazymatch c' with fun x : Set => ?f => constr:(forall x : Type, f) end in
  let _ := type of c'' in
  exact c''.
Defined.
```
It basically corresponds to https://github.com/coq/coq/issues/5996#issuecomment-338405694 Do you think that it should be like https://github.com/coq/coq/issues/5996#issuecomment-338406360 instead? I’m afraid that I don’t know anything about #5996 to have an opinion here.

**Comment by herbelin** [comment]
The current test is basically one that I cooked at some time. Indeed, it does not fully solve the original issue. I don't know what to do. @skyskimmer, would you like we add the original example to the file, with a `Fail` on the `Defined`, and move the whole file to `bugs/opened` (thus also keeping the first - current - part of the file as a working example)?

**Review by jfehrle — APPROVED** [review]
Approving the doc change. I'll leave it to others to approve the test changes.

These are small changes and somewhat related.  There's no code change that affects users.  I suggest squashing these commits after review is complete, just before merging.  

There's a tradeoff between giving many detailed commits and having a concise, readable commit history.  For code changes, we're more careful about not combining unrelated changes into a single commit.  This makes the commit history more understandable and it's easier to back out changes if unexpected issues arise afterward.

When I'm submitting a PR with changes that are complex and there may multiple rounds of review, I often put the review-related changes into a new commit, imaginatively names "Review # N" so the reviewers can easily verify the incremental changes.  And then squash them just before merging.

**Review by herbelin — APPROVED** [review]
Looks good and as expected. Shall merge tomorrow if no more comments by then.

File: doc/sphinx/addendum/extraction.rst
Comment (SkySkimmer) [inline doc/sphinx/addendum/extraction.rst:109]: This doesn't belong in this PR.

**Review by SkySkimmer — CHANGES_REQUESTED** [review]
5996 is still wrong.

**Comment by SkySkimmer** [comment]
>@SkySkimmer, would you like we add the original example to the file, with a Fail on the Defined, and move the whole file to bugs/opened (thus also keeping the first - current - part of the file as a working example)?

Sure that sounds fine

File: doc/sphinx/addendum/extraction.rst
Comment (Mbodin) [reply doc/sphinx/addendum/extraction.rst:109]: Should I split the PR again?

**Comment by Mbodin** [comment]
I added the original example. Please double-check as I’m not familiar with issue 5996.

File: test-suite/bugs/opened/bug_5996.v
Comment (SkySkimmer) [inline test-suite/bugs/opened/bug_5996.v:11]:
```suggestion
(* Workaround *)
```

File: doc/sphinx/addendum/extraction.rst
Comment (SkySkimmer) [reply doc/sphinx/addendum/extraction.rst:109]: Yes, and then squash.

**Comment by herbelin** [comment]
@skyskimmer: Do we split the PR again in two? If yes, maybe can we do it ourselves to avoid adding a burden on @Mbodin who already helped a lot?

**Comment by jfehrle** [comment]
> Do we split the PR again in two?

Meaning splitting it into 2 commits, not into 2 PRs?  One with the doc change, one with the test changes?

**Comment by Mbodin** [comment]
I’m fine with doing it, actually ☺ (I’m learning the way PR works along the way, and I actually did appreciate the learning curve ☺)
I will first try squashing into two commits (with the documentation change, and the test-suite change), then we can split the PR or leave it as-is, depending on which is better for the git history (I don’t have much opinion on which is better).

**Comment by SkySkimmer** [comment]
>I will first try squashing into two commits (with the documentation change, and the test-suite change)

No, you should split first imo

**Comment by Mbodin** [comment]
I just did the PR split ☺

**Review by SkySkimmer — APPROVED** [review]

**Pull request merged**
